### PR TITLE
Transolver++ Ada-Temp: per-point adaptive slice temperature + Rep-Slice

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current Transolver uses a **global learnable temperature** (`self.temperature`, shape `[1, heads, 1, 1]`) shared across all N=120k mesh nodes for slice weight computation. Transolver++ (arXiv 2502.02414, ICML 2025) shows that replacing this with a **per-point adaptive temperature** (Ada-Temp) prevents slice-weight homogenization — where all nodes collapse to the same slice, erasing physics distinctions.

The paper reports **62% surface-error reduction from Ada-Temp alone on aircraft datasets** and 13% average improvement across 6 PDE benchmarks. For our tandem problem, homogenization is plausible on OOD NACA6416 inputs: the 96 slices are trained overwhelmingly on NACA0012 geometries, and OOD inputs may collapse to an "average" routing that fails to represent tandem wake coupling.

**Rep-Slice** (Gumbel-Softmax reparameterization) further encourages sparse, sharp slice assignments during training by adding Gumbel noise to slice logits before softmax.

This is **distinct from WIP SCA** (frieren #2199): SCA adds a learnable diagonal scale to attention *logits after* slice tokens are computed. Ada-Temp acts *before* slice aggregation, sharpening node-to-slice routing. Complementary mechanisms.

**Source:** Transolver++, arXiv 2502.02414, ICML 2025 (same Transolver backbone as ours).

---

## Instructions

All changes are in `cfd_tandemfoil/train.py`.

### Step 1: Add argparse flags

```python
parser.add_argument('--ada_temp', action='store_true',
    help='Transolver++ per-point adaptive slice temperature')
parser.add_argument('--ada_temp_tau0', type=float, default=0.5,
    help='Base temperature for Ada-Temp (matches existing temperature init of 0.5)')
parser.add_argument('--rep_slice', action='store_true',
    help='Transolver++ Gumbel-Softmax reparameterization for sharp slice routing')
```

### Step 2: Modify `Physics_Attention_Irregular_Mesh.__init__`

This class (line ~137) takes `dim, heads, dim_head, dropout, slice_num, ...` as args. It is instantiated in `TransolverBlock`. Follow the same pattern used for `zone_temp`, `decouple_slice` — add the new booleans as constructor params and pass them through from `TransolverBlock` (which gets them from `cfg`/`args`).

In `__init__`, add:
```python
self.ada_temp_on = ada_temp        # new bool param, default False
self.rep_slice_on = rep_slice      # new bool param, default False
if self.ada_temp_on:
    # Per-point, per-head temperature offset
    # Zero-initialized so model starts IDENTICAL to baseline (offset=0)
    self.ada_temp_proj = nn.Linear(dim, heads)
    nn.init.zeros_(self.ada_temp_proj.weight)
    nn.init.zeros_(self.ada_temp_proj.bias)
```

### Step 3: Modify `Physics_Attention_Irregular_Mesh.forward`

The forward signature is `forward(self, x, spatial_bias=None, tandem_mask=None, zone_features=None)` where `x` is `[B, N, dim]`.

The existing temperature block (lines ~204–218) computes a global `temp` and uses it:
```python
temp = self.temperature            # [1, heads, 1, 1], value=0.5
# ... zone_temp and tandem_temp_offset adjustments ...
temp = temp.clamp(min=1e-4)
slice_logits = self.in_project_slice(x_mid) / temp   # [B, heads, N, slice_num]
```

**Add immediately after `temp = temp.clamp(min=1e-4)`, before the slice_logits line:**
```python
if self.ada_temp_on:
    # x is [B, N, dim]; project to per-point per-head offset
    ada_offset = self.ada_temp_proj(x)               # [B, N, heads]
    ada_offset = ada_offset.permute(0, 2, 1).unsqueeze(-1)  # [B, heads, N, 1]
    temp = (temp + ada_offset).clamp(min=0.01)       # [B, heads, N, 1] — now per-point
    # This broadcasts correctly: in_project_slice(x_mid) is [B, heads, N, slice_num]
```

**Add Rep-Slice after spatial_bias and prog_slices are applied, immediately before `slice_weights = self.softmax(slice_logits)`:**
```python
if self.rep_slice_on and self.training:
    gumbel = -torch.log(-torch.log(
        torch.empty_like(slice_logits).uniform_(1e-7, 1 - 1e-7)
    ))
    slice_logits = slice_logits + gumbel
slice_weights = self.softmax(slice_logits)
```

### Step 4: Wire up in `TransolverBlock`

Pass `ada_temp` and `rep_slice` (both default `False`) through `TransolverBlock.__init__` to the `Physics_Attention_Irregular_Mesh` constructor, following the same pattern as `zone_temp` and `decouple_slice`.

In `TransolverBlock.__init__`, read from `args`:
```python
ada_temp = getattr(args, 'ada_temp', False)
rep_slice = getattr(args, 'rep_slice', False)
```

### Step 5: Run experiments

Run 2 seeds with `--ada_temp --rep_slice`:

**Seed 42:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/ada-temp-rep-s42" \
  --wandb_group alphonse-ada-temp \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --ada_temp --ada_temp_tau0 0.5 --rep_slice \
  --seed 42
```

**Seed 73:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/ada-temp-rep-s73" \
  --wandb_group alphonse-ada-temp \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --ada_temp --ada_temp_tau0 0.5 --rep_slice \
  --seed 73
```

### Implementation notes

- **Zero-init is critical.** `ada_temp_proj` initialized to zeros so `ada_offset=0` at epoch 0 — model starts identical to baseline. Temperature only deviates as it trains.
- **VRAM impact:** One additional `[B, N, heads]` tensor per forward pass — ~7MB for B=4, N=120k, heads=8. Negligible.
- **torch.compile:** `torch.empty_like(...).uniform_()` is compile-safe.
- **Rep-Slice is training-only.** The `self.training` check skips Gumbel noise during eval and EMA passes.
- **If first run shows improvement:** Try a follow-up without `--rep_slice` (Ada-Temp only) to isolate which component drives the gain.

## Baseline (PR #2184 — DCT Freq Loss w=0.05, 2-seed avg)

| Metric | 2-seed avg | Beat threshold |
|--------|-----------|----------------|
| p_in   | 13.205    | < 13.21        |
| p_oodc | 7.816     | < 7.82         |
| **p_tan** | **28.502** | **< 28.50** |
| p_re   | 6.453     | < 6.45         |

W&B baseline runs: `6yfv5lio` (seed 42), `etepxvjc` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Transolver++ Ada-Temp + Rep-Slice

### Setup
- Ada-Temp: per-point, per-head adaptive slice temperature via zero-initialized linear projection
- Rep-Slice: Gumbel-Softmax noise added to slice logits during training for sharper routing
- Both enabled simultaneously (`--ada_temp --rep_slice`)
- 2 seeds (42, 73), full baseline flags including DCT frequency loss
- Training speed: ~5 it/s, comparable to baseline (~142 epochs reached in 3h timeout)

### Results

| Config | p_in | p_oodc | p_tan | p_re | W&B Run | Epochs |
|--------|------|--------|-------|------|---------|--------|
| **Baseline (2-seed avg)** | **13.21** | **7.82** | **28.50** | **6.45** | 6yfv5lio, etepxvjc | ~155 |
| ada_temp + rep_slice, seed 42 | 15.6 | 9.1 | 31.4 | 7.3 | `zjux1n0n` | 142 |
| ada_temp + rep_slice, seed 73 | 15.9 | 9.3 | 31.4 | 7.1 | `6t805fa9` | 141 |
| **2-seed avg** | **15.7** | **9.2** | **31.4** | **7.2** | — | ~142 |

### Analysis

**Negative result.** All metrics are significantly worse than baseline:
- p_in: +2.5 (15.7 vs 13.2, +19% regression)
- p_oodc: +1.4 (9.2 vs 7.8, +18% regression)
- p_tan: +2.9 (31.4 vs 28.5, +10% regression)
- p_re: +0.8 (7.2 vs 6.5, +12% regression)

**Possible causes:**

1. **Ada-Temp + Rep-Slice combined may be too aggressive.** The Gumbel noise from Rep-Slice makes slice logits stochastic, while Ada-Temp simultaneously modifies the temperature. The combination may destabilize slice routing in early training, preventing convergence to good slice assignments.

2. **Interaction with existing temperature mechanisms.** The model already has `tandem_temp_offset` and `zone_temp_proj` for adaptive temperature control. Adding a third per-point temperature offset on top of these may be over-parameterized — the 3 competing temperature mechanisms may fight each other.

3. **Slightly fewer epochs** (142 vs ~155 baseline) may contribute, though this ~8% reduction alone doesn't explain the 18-19% regression.

4. **Different training dynamics.** The Transolver++ paper uses standard Adam with different training schedules. Our pipeline uses Lion optimizer + PCGrad + EMA + heavy augmentation, which may not pair well with the Gumbel noise from Rep-Slice.

### Suggested Follow-ups
- Try **Ada-Temp only** (without Rep-Slice) to isolate whether the Gumbel noise is the problem
- Try **Rep-Slice only** to check if the temperature mechanism causes the regression
- If both fail individually, the Transolver++ improvements may not transfer to our specific training setup

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2206 because ADVISOR: Closing — clear dead end. All metrics regressed significantly (p_tan +10%, p_in +19%, p_oodc +18%, p_re +12%).

**Root cause analysis:**
1. **Over-parameterized temperature control.** The model already has tandem_temp_offset and zone_temp_proj for adaptive temperature. Adding a 3rd per-point temperature projection creates competing mechanisms that fight each other rather than cooperate.
2. **Rep-Slice Gumbel noise is incompatible with our training setup.** Our pipeline uses Lion optimizer + PCGrad gradient surgery + EMA — the stochastic Gumbel noise from Rep-Slice adds variance that PCGrad cannot surgically correct, destabilizing convergence.
3. **Transolver++ paper used fundamentally different training dynamics** (standard Adam, no PCGrad, no EMA, different schedules). The improvements don't transfer to our heavily-tuned pipeline.

**On your suggested follow-ups:** While isolating Ada-Temp vs Rep-Slice is scientifically clean, the 10-19% regression magnitude suggests neither component alone is likely to help given our existing temperature mechanisms. The existing zone_temp_proj already provides per-point adaptive temperature conditioning. Direction exhausted.

Good analysis of the failure modes — your observation about the interaction with existing temperature mechanisms was spot on. Moving you to a fresh direction..
